x/batching/keeper/evidence.go

```diff
@@ -0,0 +1,217 @@
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/sedaprotocol/seda-chain/app/utils"
+	"github.com/sedaprotocol/seda-chain/x/batching/types"
+)
+
+func NewBatchDoubleSignHandler(keeper Keeper) func(ctx context.Context, evidence exported.Evidence) error {
+	return func(ctx context.Context, evidence exported.Evidence) error {
+		return keeper.handleEvidence(ctx, evidence.(*types.BatchDoubleSign))
+	}
+}
+
+func (k *Keeper) handleEvidence(ctx context.Context, evidence *types.BatchDoubleSign) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	sdkCtx.Logger().Info("received batch double sign evidence", "batch height", evidence.BatchHeight, "operator address", evidence.OperatorAddress, "result root", evidence.DataResultRoot, "validator root", evidence.ValidatorRoot, "proving metadata hash", evidence.ProvingMetadataHash, "proving scheme index", evidence.ProvingSchemeIndex)
+
+	// Validate that a batch exists for the given batch height.
+	batch, err := k.GetBatchByBatchNumber(ctx, evidence.BatchHeight)
+	if err != nil {
+		return err
+	}
+
+	if batch.BlockHeight != evidence.BlockHeight {
+		return fmt.Errorf("batch height mismatch. Expected: %d, Got: %d", batch.BlockHeight, evidence.BlockHeight)
+	}
+
+	// Validate the signed batch is different from what was recorded on chain.
+	fraudulentBatchID, err := evidence.GetBatchID()
+	if err != nil {
+		return err
+	}
+
+	if bytes.Equal(batch.BatchId, fraudulentBatchID) {
+		return fmt.Errorf("batch IDs are the same")
+	}
+
+	// Currently we only support secp256k1 signatures so no need to check which proving scheme the validator used.
+	signatureAddr, err := k.recoverEthAddressFromSecp256k1Signature(fraudulentBatchID, evidence.Signature)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the validator entry from the previous batch, as they might have changed their public key in the
+	// fraudulent batch.
+	validatorEthAddr, err := k.getEthAddressForBatch(ctx, evidence.BatchHeight-1, evidence.OperatorAddress)
+	if err != nil {
+		return err
+	}
+
+	// If the recovered address matches the validator entry they have committed a double sign.
+	if !bytes.Equal(validatorEthAddr, signatureAddr) {
+		return fmt.Errorf("recovered address does not match validator entry. Recorded: %s, Got: %s", hex.EncodeToString(validatorEthAddr), hex.EncodeToString(signatureAddr))
+	}
+
+	sdkCtx.Logger().Info("confirmed double batch sign", "validator", evidence.OperatorAddress, "batch height", evidence.BatchHeight, "block height", evidence.BlockHeight)
+
+	// Reject evidence if the double-sign is too old. Evidence is considered stale
+	// if the difference in number of blocks is greater than the allowed
+	// parameter defined.
+	infractionHeight := evidence.GetHeight()
+	ageBlocks := sdkCtx.BlockHeader().Height - infractionHeight
+	cp := sdkCtx.ConsensusParams()
+	if cp.Evidence != nil {
+		if ageBlocks > cp.Evidence.MaxAgeNumBlocks {
```

Comment:
Technically this will never happen because the historical info look up will fail after 10,000 blocks (`historical_entries` parameter) while this `max_age_num_blocks` param is 100,000, right?

Reply:
Hmm, that's a pity. The block equivocation handler relies on the power as it is submitted on the evidence, but we have no way of really verifying the power if we let a user submit it in the TX.

I suppose we can drop this check and replace it with the historical entries from the staking keeper?

Reply:
Found that CometBFT uses the state store to load the validator set at the time of infraction (see https://github.com/cometbft/cometbft/blob/9722b6dfa1d0da3c3d4f73c1c779479971959d94/evidence/verify.go#L51) but I wasn't able to figure out whether this store is accessible from the application layer. I don't think it is.